src/main/java/net/dv8tion/jda/api/JDA.java

```diff
@@ -378,6 +379,71 @@ default JDA awaitReady() throws InterruptedException
         return awaitStatus(Status.CONNECTED);
     }
 
+    /**
+     * Blocks the current thread until {@link #getStatus()} returns {@link Status#SHUTDOWN}.
+     * <br>This can be useful in certain situations like disabling class loading.
+     *
+     * <p>Note that shutdown time depends on the length of the rate-limit queue.
+     * You can use {@link #shutdownNow()} to cancel all pending requests and immediately shutdown.
+     *
+     * <p><b>Example</b>
+     * <pre>{@code
+     * jda.shutdown();
+     * // Allow at most 10 seconds for remaining requests to finish
+     * if (!jda.awaitShutdown(Duration.ofSeconds(10))) {
+     *     jda.shutdownNow(); // Cancel all remaining requests
+     *     jda.awaitShutdown(); // Wait until shutdown is complete (indefinitely)
+     * }
+     * }</pre>
+     *
+     * <p><b>This will not implicitly call {@code shutdown()}, you are responsible to ensure that the shutdown process has started.</b>
+     *
+     * @param  timeout
+     *         The maximum time to wait, or {@link Duration#ZERO} to wait indefinitely
+     *
+     * @throws IllegalArgumentException
+     *         If the provided timeout is null
+     * @throws InterruptedException
+     *         If the current thread is interrupted while waiting
+     *
+     * @return False, if the timeout has elapsed before the shutdown has completed, true otherwise.
+     */
+    @CheckReturnValue
+    boolean awaitShutdown(@Nonnull Duration timeout) throws InterruptedException;
```

Comment:
We should consider accepting `long, TemporalUnit` as it is what a lot of our apis already accept.